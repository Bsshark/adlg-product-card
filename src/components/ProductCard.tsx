import React, { CSSProperties, createContext } from "react";

import { useProduct } from "../hooks/useProduct";
import {
	InitialValues,
	Product,
	ProductCardHandlers,
	ProductContextProps,
	onChangeArgs,
} from "../interfaces/ProductInterfaces";
import styles from "../styles/styles.module.css";

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export interface Props {
	product: Product;
	/* children?: ReactElement | ReactElement[]; */
	children: (args: ProductCardHandlers) => JSX.Element;
	className?: string;
	style?: CSSProperties;
	onChange?: (args: onChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({
	children,
	product,
	className,
	style,
	onChange,
	value,
	initialValues,
}: Props) => {
	const { counter, increaseBy, maxQuantity, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues,
		});
	return (
		<Provider
			value={{
				counter,
				increaseBy,
				product,
				maxQuantity,
			}}
		>
			<div className={`${styles.productCard} ${className}`} style={style}>
				{children({
					count: counter,
					isMaxCountReached,
					maxCount: initialValues?.maxQuantity,
					product,
					increaseBy,
					reset,
				})}
			</div>
		</Provider>
	);
};

/* export const ProductButtons = ({
  counter,
  handleClick,
  className,
}: ProductButtons) => {
  return (
    <div className={className}>
      <ButtonComponent
        className={styles.buttonMinus}
        handleClick={handleClick}
        increaseValue={-1}
        char={"-"}
      />

      <div className={styles.countLabel}> {counter} </div>
      <ButtonComponent
        className={styles.buttonAdd}
        handleClick={handleClick}
        increaseValue={1}
        char={"+"}
      />
    </div>
  );
}; */

/* export const ButtonComponent = ({
  handleClick,
  increaseValue,
  className,
  char,
}: ButtonProduct) => {
  return (
    <button className={className} onClick={() => handleClick(increaseValue)}>
      {char}
    </button>
  );
}; */

/* interface ProductButtons {
  counter: number;
  increaseBy: (value: number) => void;
}

interface ButtonProduct {
  handleClick: (n: number) => void;
  increaseValue: number;
  className: string;
  char: string;
} */
