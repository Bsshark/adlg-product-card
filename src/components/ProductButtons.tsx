import React, { CSSProperties, useCallback, useContext } from "react";
import styles from "../styles/styles.module.css";
import { ProductContext } from "./ProductCard";

export interface Props {
	className?: string;
	style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
	const { increaseBy, counter, maxQuantity } = useContext(ProductContext);

	// TODO useCallback isMaxReached [quantity, maxQuantity]
	const isMaxReached = useCallback(() => {
		return maxQuantity === counter ? true : false;
	}, [counter, maxQuantity]);
	const isMinReached = useCallback(() => {
		return counter === 0 ? true : false;
	}, [counter]);

	return (
		<div className={`${styles.buttonsContainer} ${className}`} style={style}>
			<button
				className={`${styles.buttonMinus} ${
					isMinReached() ? styles.disabled : ""
				}`}
				onClick={() => increaseBy(-1)}
			>
				-
			</button>

			<div className={styles.countLabel}> {counter} </div>
			<button
				className={`${styles.buttonAdd} ${
					isMaxReached() ? styles.disabled : ""
				}`}
				onClick={() => increaseBy(1)}
			>
				+
			</button>
		</div>
	);
};
