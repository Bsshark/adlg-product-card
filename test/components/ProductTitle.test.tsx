import renderer from 'react-test-renderer';
import { ProductTitle } from '../../src/components/ProductTitle';
import React from 'react';
import { product1 } from '../data/productos';
import { ProductCard } from '../../src/components/ProductCard';

describe('Pruebas en Product Title', () => {
  it('debe mostrar el componente correctamente con el titulo personalizado ', () => {
    const customTitle = 'Custom Product Title';
    const wrapper = renderer.create(<ProductTitle title={customTitle} />);

    expect(wrapper).toMatchSnapshot();
  });
  it('debe mostrar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
