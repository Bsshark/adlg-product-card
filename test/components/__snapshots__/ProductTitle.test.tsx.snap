// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Product Title debe mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className=" productDescription  undefined"
  >
    Coffe Mug - No Image
  </span>
</div>
`;

exports[`Pruebas en Product Title debe mostrar el componente correctamente con el titulo personalizado  1`] = `
<span
  className=" productDescription  undefined"
>
  Custom Product Title
</span>
`;
